### PR TITLE
[risk=low][RW-7586] Keep going after failing to set Workflow Runner roles

### PR DESCRIPTION
**Pull request by jmthibault79**
Tested locally by itself and with #6519 
Workspace Share and Duplication continue to work, with expected API JSON output

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [x] I have added explanatory comments where the logic is not obvious
- [x] I have run and tested this change locally, and my testing process is described here
- [x] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [x] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [x] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [x] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [x] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/iam/IamService.java [hunk 1]
```diff
@@ -4,26 +4,29 @@
 
 public interface IamService {
   /**
-   * Grants user permissions to run Google lifescience jobs to user in currenet context.
+   * Grants user permissions to run Google Life Sciences jobs to the user in the current context.
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/iam/IamService.java:7]: did a grammar/spelling pass of this file while I was here

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 2]
```diff
@@ -602,6 +602,8 @@ private void validateWorkspaceApiModel(Workspace workspace) {
     }
 
     dbWorkspace = workspaceDao.saveWithLastModified(dbWorkspace);
+    final Workspace savedWorkspace = workspaceMapper.toApiWorkspace(dbWorkspace, toFcWorkspace);
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java:605]: moved up from below, so we have a `CloneWorkspaceResponse` to fill in

File: api/src/main/java/org/pmiops/workbench/iam/IamServiceImpl.java [hunk 3]
```diff
@@ -72,40 +84,42 @@ private void grantServiceAccountUserRole(String googleProject, String petService
...
+
+    for (String userEmail : userEmails) {
+      try {
         String petServiceAccountName =
-            getOrCreatePetServiceAccount(googleProject, googleApiAsImpersonatedUser);
+            getOrCreatePetServiceAccountUsingImpersonation(googleProject, userEmail);
         petServiceAccountsToGrantPermission.add(petServiceAccountName);
         grantServiceAccountUserRole(googleProject, petServiceAccountName);
+      } catch (Exception e) {
+        petServiceAccountFailures.add(userEmail);
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/iam/IamServiceImpl.java:99]: key difference: collects individual email failures but keeps going to try others

File: api/src/main/java/org/pmiops/workbench/iam/IamServiceImpl.java [hunk 4]
```diff
@@ -24,40 +27,64 @@
...
+   * not yet exist.
+   */
+  private String getOrCreatePetServiceAccountAsCurrentUser(String googleProject) {
+    return getOrCreatePetServiceAccount(googleProject, endUserGoogleApiProvider.get());
+  }
+
+  /**
+   * Gets a Terra pet service account from SAM as the given user using impersonation, if possible.
+   *
+   * <p>If the user has not yet accepted the latest Terms of Service, impersonation will not be
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/iam/IamServiceImpl.java:69]: Inspired by conversations in https://docs.google.com/document/d/1tVA-Pv6R65N2fJQw74-StgEmYAKkZLjj5WyeE9R-DGs/edit#

File: api/src/main/java/org/pmiops/workbench/iam/IamServiceImpl.java [hunk 5]
```diff
@@ -24,40 +27,64 @@
...
-  private String getOrCreatePetServiceAccount(String googleProject, GoogleApi googleApi) {
-    return samRetryHandler.run(
-        (context) -> {
-          return googleApi.getPetServiceAccount(googleProject);
-        });
+  /**
+   * Gets a Terra pet service account from SAM as the current user. SAM will create one if one does
+   * not yet exist.
+   */
+  private String getOrCreatePetServiceAccountAsCurrentUser(String googleProject) {
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/iam/IamServiceImpl.java:62]: not strictly necessary - it's a convenience method to make it explicit that this is the current-user version

**Review by evrii — APPROVED** [review]
Looks good to me. If you get a chance, do you mind walking me through this again? Want to make sure I fully understand the process.

File: api/src/main/java/org/pmiops/workbench/iam/IamServiceImpl.java (hunk 5 above)
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/iam/IamServiceImpl.java:62]: (discussed offline) The generic method takes any credentials.  This method passes the End User GoogleApi to that one, so the current user's credentials are used to call the API.

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 6]
```diff
@@ -704,15 +712,18 @@ private FirecloudWorkspaceId getFcBillingProject(DbAccessTier accessTier, Worksp
                               || WorkspaceAccessLevel.WRITER.equals(u.getRole()))
                           && !finalWorkflowUsers.contains(u.getEmail())
                           && !u.getEmail().equals(userProvider.get().getUsername()))
-              .map(u -> u.getEmail())
+              .map(UserRole::getEmail)
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java:715]: yes

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 7]
```diff
@@ -704,15 +712,18 @@ private FirecloudWorkspaceId getFcBillingProject(DbAccessTier accessTier, Worksp
...
-          dbWorkspace.getGoogleProject(), userLostPermission);
+      List<String> failedRevocations =
+          iamService.revokeWorkflowRunnerRoleForUsers(
+              dbWorkspace.getGoogleProject(), userLostPermission);
+      if (!failedRevocations.isEmpty()) {
+        resp.setFailedWorkflowRevocations(failedRevocations);
+      }
     }
 
     // This automatically enforces the "canShare" permission.
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java:725]: Unrelated, existing issue: re-reviewing this code, I believe this is a bug: https://precisionmedicineinitiative.atlassian.net/browse/RW-8158

**Review by calbach — APPROVED** [review]
Looks good thanks.

Out of curiosity, did we confirm that sharing to an "un-TOS'd" user works in general? I guess this would be a general issue in Terra (not just AoU) if that didn't work.

**Comment by jmthibault79** [comment]
Good question!  I just un-TOS'd a user (thanks to the new `-X DELETE https://sam.dsde-dev.broadinstitute.org/register/user/v1/termsofservice`) and I was still able to share with that user.

**Pull request merged**